File: cses/02_sorting_searching/2163_josephus-problem-ii/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    k = int(data[1])  # số đứa bị bỏ qua mỗi lượt

    # Order-statistics segment tree đếm số phần tử còn tồn tại trên [1..n]
    # Kích thước lá: lũy thừa 2 nhỏ nhất >= n
    size = 1
    while size < n:
        size <<= 1
    tree = [0] * (2 * size)
    for i in range(n):
        tree[size + i] = 1                   # mỗi đứa ban đầu còn tồn tại -> lá = 1
    for i in range(size - 1, 0, -1):
        tree[i] = tree[2 * i] + tree[2 * i + 1]

    res = [0] * n
    pos = 0                                  # điểm bắt đầu đếm (0-based) trong các phần tử còn lại
    remaining = n
    for out in range(n):
        idx = (pos + k) % remaining          # chỉ số 0-based của đứa trẻ bị loại
        target = idx + 1                     # tìm phần tử thứ target (1-based) còn tồn tại

        # Đi xuống một lần: vừa tìm phần tử thứ target vừa xóa nó dọc đường
        node = 1
        tree[1] -= 1
        while node < size:
            left = node + node
            lc = tree[left]
            if lc >= target:
                node = left                  # đủ phần tử bên trái -> rẽ trái
            else:
                target -= lc                 # bỏ qua nhánh trái -> rẽ phải
                node = left + 1
            tree[node] -= 1

        res[out] = node - size + 1           # số hiệu đứa trẻ (1-based)
        pos = idx                            # đứa ngay sau chiếm đúng chỉ số idx trong mảng mới
        remaining -= 1

    sys.stdout.write(' '.join(map(str, res)) + '\n')


main()

File: cses/02_sorting_searching/2163_josephus-problem-ii/solution.cpp
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n;
    long long skipped;  // số đứa bị bỏ qua mỗi lượt (k), có thể tới 10^9
    std::cin >> n >> skipped;

    // Kích thước lá của segment tree: lũy thừa 2 nhỏ nhất >= n
    int size = 1;
    while (size < n) {
        size *= 2;
    }

    // Order-statistics segment tree: mỗi lá 1..n = 1 (còn tồn tại), node trong = tổng hai con
    std::vector<int> tree(2 * size, 0);
    for (int i = 0; i < n; ++i) {
        tree[size + i] = 1;
    }
    for (int node = size - 1; node >= 1; --node) {
        tree[node] = tree[2 * node] + tree[2 * node + 1];
    }

    long long position = 0;  // chỉ số 0-based điểm bắt đầu đếm trong các phần tử còn lại
    for (int remaining = n; remaining >= 1; --remaining) {
        // Chỉ số 0-based của đứa bị loại lượt này, rồi đổi sang thứ hạng 1-based
        const long long removed_index = (position + skipped) % remaining;
        int target = static_cast<int>(removed_index) + 1;

        // Đi xuống một lần: vừa tìm phần tử thứ target còn tồn tại vừa xóa nó dọc đường
        int node = 1;
        --tree[node];
        while (node < size) {
            const int left = 2 * node;
            if (tree[left] >= target) {
                node = left;                 // đủ phần tử bên trái -> rẽ trái
            } else {
                target -= tree[left];        // bỏ qua nhánh trái -> rẽ phải
                node = left + 1;
            }
            --tree[node];
        }

        // Ghi số hiệu đứa vừa bị loại (node - size + 1 là lá 1-based)
        if (remaining < n) {
            std::cout << ' ';
        }
        std::cout << node - size + 1;

        // Đứa ngay sau chiếm đúng chỉ số removed_index trong mảng còn lại mới
        position = removed_index;
    }
    std::cout << '\n';

    return 0;
}
